packages/core/src/lib/normalize.unit.test.ts

```diff
@@ -0,0 +1,112 @@
+import { join } from 'node:path';
+import { describe, expect } from 'vitest';
+import { AuditOutput, Issue } from '@code-pushup/models';
+import { normalizeAuditOutputs, normalizeIssue } from './normalize';
+
+vi.mock('@code-pushup/utils', async () => {
+  const actual = await vi.importActual('@code-pushup/utils');
+
+  return {
+    ...actual,
+    gitRoot: vi.fn().mockResolvedValue('/User/code-pushup'),
+  };
+});
+
+describe('normalizeAuditOutputs', () => {
+  it('should forward empty auditOutputs', async () => {
+    await expect(normalizeAuditOutputs([])).resolves.toEqual([]);
+  });
+
+  it('should forward audits without details', async () => {
+    await expect(
+      normalizeAuditOutputs([{ slug: 'no-any' } as AuditOutput]),
+    ).resolves.toEqual([{ slug: 'no-any' }]);
+  });
+
+  it('should forward audit details with table', async () => {
+    const outputs = await normalizeAuditOutputs([
+      { details: { table: {} } } as unknown as AuditOutput,
+    ]);
+    expect(outputs.at(0)).toStrictEqual({ details: { table: {} } });
+  });
+
+  it('should forward audit details without issues', async () => {
+    const outputs = await normalizeAuditOutputs([
+      { details: { issues: undefined } } as unknown as AuditOutput,
+    ]);
+    expect(outputs.at(0)).toStrictEqual({ details: { issues: undefined } });
+  });
+
+  it('should forward audit details with empty issues', async () => {
+    const outputs = await normalizeAuditOutputs([
+      { details: { issues: [] } } as unknown as AuditOutput,
+    ]);
+    expect(outputs.at(0)).toStrictEqual({ details: { issues: [] } });
+  });
+
+  it('should forward audit details with issues and all undefined source', async () => {
+    const issues = [
+      { source: undefined },
+      { source: undefined },
+      { source: undefined },
+    ];
+    const outputs = await normalizeAuditOutputs([
+      { details: { issues } } as unknown as AuditOutput,
+    ]);
+    expect(outputs.at(0)?.details?.issues).toBe(issues);
+  });
+
+  it('should clone audit details with issues that have source specified', async () => {
+    const issues = [
+      { source: undefined },
+      { source: { file: 'index.js' } },
```

Comment:
You're testing a function whose primary purpose is to transform issue source paths. There are a ton of tests concerned with when it does and doesn't clone audit objects, which is fine, but this optimization is a secondary concern. There are no tests concerned with its primary purpose - finding all non-relative source paths and making them relative. 

You can test this by passing in source paths which aren't already relative, e.g.:

```suggestion
      { source: { file: join(MEMFS_VOLUME, 'index.js') } },
```

Reply:
I did not fully understand the path logic so i implemented unit tests for `formatGitPath`.

For the `normalizeIssue` function I would write this:

```ts
it('should normalize filepath in issue if source file is given', () => {
    vi.spyOn(process, 'cwd').mockReturnValue('/Users/user/Projects/myProject');
    const path = '/Users/user/Projects/myProject/utils/index.js';
    const gitRoot = '/Users/user/Projects/myProject';
    expect(
      normalizeIssue(
        {
          message: 'file too big',
          severity: 'error',
          source: {
            file: path,
          },
        },
        gitRoot,
      ),
    ).toEqual({
      message: 'file too big',
      severity: 'error',
      source: {
        file: 'utils/index.js',
      },
    });
  });
```

Also I saw that the there is another helper `toGitPath` that could be used in `normalizeIssue` only and removes the use of any git API in `normalizeAuditOutputs`. But I'm not sure if calling gitRoot has perf degradations when calling it many times. ATM there is only one call for all issues in an audit.

Reply:
Cool, I like both unit tests a lot :heart_eyes: